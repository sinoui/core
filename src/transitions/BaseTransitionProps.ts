import type {
  TransitionProps,
  TransitionActions,
} from 'react-transition-group/Transition';

type TransitionKeys =
  | 'in'
  | 'mountOnEnter'
  | 'unmountOnExit'
  | 'onEnter'
  | 'onEntering'
  | 'onEntered'
  | 'onExit'
  | 'onExiting'
  | 'onExited'
  | 'timeout'
  | 'addEndListener';

/**
 * 基础的过度组件属性
 */
export default interface BaseTransitionProps
  extends TransitionActions,
    Partial<Pick<TransitionProps, TransitionKeys>> {
  style?: React.CSSProperties;
  className?: string;
  children: React.ReactElement;
}
