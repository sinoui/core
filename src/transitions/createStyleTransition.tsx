import React from 'react';
import { CSSTransition } from 'react-transition-group';
import { transitions } from '@sinoui/theme';
import { getDuration, getEasing } from './transitionUtils';

export interface StyleTransitionPropsType {
  appear?: boolean;
  in?: boolean;
  /**
   * 过渡效果的方向
   *
   * @type {('left' | 'right')}
   */
  direction?: 'left' | 'right' | 'up' | 'down';
  /**
   * 子节点
   *
   * @type {Node}
   */
  children?: React.ReactNode;
  /**
   * 过渡时长。'auto'表示会根据元素高度自动计算过渡时长
   *
   * @type {(number
   *     | {
   *         enter: number,
   *         exit: number,
   *       })}
   */
  timeout?:
    | 'auto'
    | number
    | {
        enter: number;
        exit: number;
      };
  /**
   * 指定过渡时应用的css类名。
   */
  transitionClasses?:
    | string
    | {
        appear?: string;
        appearActive?: string;
        enter?: string;
        enterActive?: string;
        exit?: string;
        exitActive?: string;
      };
  easing?: string;
  onEnter?: (node: HTMLElement, isAccept: boolean) => void;
  onEntering?: (node: HTMLElement, isAccept: boolean) => void;
  onEntered?: (node: HTMLElement, isAccept: boolean) => void;
  onExit?: (node: HTMLElement) => void;
  onExiting?: (node: HTMLElement) => void;
  onExited?: (node: HTMLElement) => void;
}

export interface StyleTransitionOptions {
  enterStyle: (
    transition: TransitionInterface,
    node: HTMLElement,
    isAppearing: boolean,
  ) => React.CSSProperties;
  enterActiveStyle?: (
    transition: TransitionInterface,
    node: HTMLElement,
    isAppearing: boolean,
  ) => React.CSSProperties;
  enterToStyle: (
    transition: TransitionInterface,
    node: HTMLElement,
    isAppearing: boolean,
  ) => React.CSSProperties;
  exitStyle: (
    transition: TransitionInterface,
    node: HTMLElement,
  ) => React.CSSProperties;
  exitActiveStyle?: (
    transition: TransitionInterface,
    node: HTMLElement,
  ) => React.CSSProperties;
  exitToStyle: (
    transition: TransitionInterface,
    node: HTMLElement,
  ) => React.CSSProperties;
  transitionProps?: string[];
  duration?: {
    enter: number;
    exit: number;
  };
  easing?: {
    enter: string;
    exit: string;
  };
}

export interface TransitionInterface {
  props: StyleTransitionPropsType;
  getTimeout: (
    e: HTMLElement,
  ) => {
    enter: number;
    exit: number;
  };
}

/**
 * 创建通过css样式定义过渡效果的过渡组件。
 *
 * 元素入场过渡： 先应用`enterStyle`，一帧（16ms）之后，应用`enterActiveStyle`和`enterToStyle`。
 *
 * 元素出场过渡： 先应用`exitStyle`，一帧（16ms）之后，应用`exitActiveStyle`和`enterToStyle`。
 *
 */
export default function createStyleTransition<Props extends {}>(
  options: StyleTransitionOptions,
  defaultProps?: Props & StyleTransitionPropsType,
): React.ComponentType<Props & StyleTransitionPropsType> {
  return class StyleTransition
    extends React.Component<Props & StyleTransitionPropsType>
    implements TransitionInterface {
    public static defaultProps = defaultProps;

    constructor(props: Props & StyleTransitionPropsType) {
      super(props);

      this.handleEnter = this.handleEnter.bind(this);
      this.handleEntering = this.handleEntering.bind(this);
      this.handleExit = this.handleExit.bind(this);
      this.handleExiting = this.handleExiting.bind(this);
      this.handleTransitionEnd = this.handleTransitionEnd.bind(this);
    }

    get easing(): {
      enter: string;
      exit: string;
    } {
      return getEasing(this.props.easing, options.easing);
    }

    public getTimeout(
      node: HTMLElement,
    ): {
      enter: number;
      exit: number;
    } {
      const timeout:
        | 'auto'
        | number
        | {
            enter: number;
            exit: number;
          } = this.props.timeout || options.duration;
      if (timeout === 'auto') {
        const duration = transitions.getAutoHeightDuration(node.clientHeight);
        return {
          enter: duration,
          exit: duration,
        };
      }
      return getDuration(timeout, options.duration);
    }

    /**
     * 处理过渡的enter事件
     *
     * @param {HTMLElement} node 应用过渡效果的节点
     * @param {boolean} isAppearing 表示是否是刚创建节点时机。
     *  如果`HorizontalSlide`组件设置了`appear`属性，则会在刚创建节点时，
     *  调用此方法，且isAppearing为true；否则不会调用此方法。
     */
    public handleEnter(node: HTMLElement, isAppearing: boolean) {
      if (this.props.onEnter) {
        this.props.onEnter(node, isAppearing);
      }

      const enterStyle = {
        transition: 'none',
        ...options.enterStyle(this, node, isAppearing),
      };
      Object.assign(node.style, enterStyle);
    }

    /**
     * 处理过渡的entering事件
     *
     * @param {HTMLElement} node 应用过渡效果的节点
     * @param {boolean} isAppearing 表示是否是刚创建节点时机。
     *  如果`HorizontalSlide`组件设置了`appear`属性，则会在刚创建节点时，
     *  调用此方法，且isAppearing为true；否则不会调用此方法。
     */
    public handleEntering(node: HTMLElement, isAppearing: boolean) {
      const activeStyle = options.enterActiveStyle
        ? options.enterActiveStyle(this, node, isAppearing)
        : {};
      if (!activeStyle.transition) {
        activeStyle.transition = transitions.create(options.transitionProps, {
          duration: this.getTimeout(node).enter,
          easing: this.easing.enter,
        });
      }

      const enterActiveStyle = {
        ...activeStyle,
        ...options.enterToStyle(this, node, isAppearing),
      };

      Object.assign(node.style, enterActiveStyle);

      if (this.props.onEntering) {
        this.props.onEntering(node, isAppearing);
      }
    }

    /**
     * 处理过渡的exit事件
     *
     * @param {HTMLElement} node 应用过渡效果的节点
     * @param {boolean} isAppearing 表示是否是刚创建节点时机。
     *  如果`HorizontalSlide`组件设置了`appear`属性，则会在刚创建节点时，
     *  调用此方法，且isAppearing为true；否则不会调用此方法。
     */
    public handleExit(node: HTMLElement) {
      const exitStyle = {
        transition: 'none',
        ...options.exitStyle(this, node),
      };

      Object.assign(node.style, exitStyle);

      if (this.props.onExit) {
        this.props.onExit(node);
      }
    }

    /**
     * 处理过渡的exiting事件
     *
     * @param {HTMLElement} node 应用过渡效果的节点
     */
    public handleExiting(node: HTMLElement) {
      const exitActiveStyle = options.exitActiveStyle
        ? options.exitActiveStyle(this, node)
        : {};
      if (!exitActiveStyle.transition) {
        exitActiveStyle.transition = transitions.create(
          options.transitionProps,
          {
            duration: this.getTimeout(node).exit,
            easing: this.easing.exit,
          },
        );
      }
      const style = {
        ...exitActiveStyle,
        ...options.exitToStyle(this, node),
      };

      Object.assign(node.style, style);

      if (this.props.onExiting) {
        this.props.onExiting(node);
      }
    }

    /**
     * 过渡结束监听器。
     *
     * @param {HTMLElement} node 过渡元素
     * @param {() => void} next 通知`Transition`组件过渡结束的方法。
     */
    public handleTransitionEnd(node: HTMLElement, next: () => void) {
      if (this.props.timeout === 'auto') {
        setTimeout(next, this.getTimeout(node).enter);
      }
    }

    public render() {
      const { transitionClasses, direction, timeout, ...props } = this
        .props as StyleTransitionPropsType;
      return (
        <CSSTransition
          {...props}
          onEnter={this.handleEnter}
          onEntering={this.handleEntering}
          onExit={this.handleExit}
          onExiting={this.handleExiting}
          timeout={
            timeout === 'auto' ? null : getDuration(timeout, options.duration)
          }
          classNames={transitionClasses || ''}
          addEndListener={this.handleTransitionEnd}
        />
      );
    }
  };
}
