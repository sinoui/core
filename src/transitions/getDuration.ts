import { transitions } from '@sinoui/theme';
import type BaseTransitionProps from './BaseTransitionProps';

export type TransitionTimeout = BaseTransitionProps['timeout'] | 'auto';

/**
 * 获取过渡时长
 *
 * @param timeout 时长
 * @param node 节点
 * @param mode 模式
 */
const getDuration = (
  timeout: TransitionTimeout,
  status: 'enter' | 'exit',
  node?: HTMLElement | null,
) => {
  if (!timeout) {
    return 0;
  }
  if (timeout === 'auto') {
    return node ? transitions.getAutoHeightDuration(node.clientHeight) : 0;
  }
  if (typeof timeout === 'number') {
    return timeout;
  }
  return timeout[status] ?? 0;
};

export default getDuration;
