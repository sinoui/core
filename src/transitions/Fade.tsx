import { transitions } from '@sinoui/theme';
import createStyleTransition from './createStyleTransition';

const { duration, easing } = transitions;

// 默认的过渡时长
const TRANSITION_DURATION = {
  enter: duration.enteringScreen,
  exit: duration.leavingScreen,
};

/**
 * 淡入淡出效果。内部使用
 * [react-transition-group/CSSTransition](https://reactcommunity.org/react-transition-group/#CSSTransition)实现过渡效果。
 */
const Fade = createStyleTransition({
  enterStyle: () => ({
    opacity: 0,
  }),
  enterToStyle: () => ({
    opacity: 1,
  }),
  exitStyle: () => ({
    opacity: 1,
  }),
  exitToStyle: () => ({
    opacity: 0,
  }),
  transitionProps: ['opacity'],
  duration: TRANSITION_DURATION,
  easing: {
    enter: easing.easeInOut,
    exit: easing.easeInOut,
  },
});

export default Fade;
