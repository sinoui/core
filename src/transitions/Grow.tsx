import { transitions } from '@sinoui/theme';
import createStyleTransition, {
  TransitionInterface,
} from './createStyleTransition';

function getScale(value: number) {
  return `scale(${value}, ${value ** 2})`;
}

export function getEnterTransition(
  instance: TransitionInterface,
  node: HTMLElement,
) {
  const duration = instance.getTimeout(node);
  const enterDuration = duration ? duration.enter : 300;
  return [
    transitions.create('opacity', {
      duration: enterDuration,
    }),
    transitions.create('transform', {
      duration: enterDuration * 0.666,
    }),
  ].join(',');
}

export function getExitTransition(
  instance: TransitionInterface,
  node: HTMLElement,
) {
  const duration = instance.getTimeout(node);
  const exitDuration = duration ? duration.exit : 300;

  return [
    transitions.create('opacity', {
      duration: exitDuration,
    }),
    transitions.create('transform', {
      duration: exitDuration * 0.666,
      delay: exitDuration * 0.333,
    }),
  ].join(',');
}

const defaultProps: {
  timeout?:
    | 'auto'
    | number
    | {
        enter: number;
        exit: number;
      };
} = {
  timeout: 'auto',
};

/**
 * 增大、缩小的过渡效果。
 *
 * ```javascript
 * import React from 'react';
 * import Grow from 'sinoui-components/transitions/Grow';
 * import { TransitionInterface } from './createStyleTransition';
 *
 * class GrowDemo extends React.Component {
 *  state: {
 *    in: false,
 *  };
 *
 *  handleButtonClick = () => this.setState(prevState => ({ !prevState.in}));
 *
 *  render() {
 *    return <div>
 *      <button onClick={this.handleButtonClick}>切换Grow动画</button>
 *      <Grow in={this.state.in}>这是应用动画的内容</Grow>
 *    </div>
 *  }
 * }
 * ```
 *
 * 增大：
 *
 * * opacity: 0 -> 1
 * * scaleX: 0.75 -> 1
 * * scaleY: 0.5625 -> 1
 *
 * 缩小：
 *
 * * opacity: 1 -> 0
 * * scaleX: 1 -> 0.75
 * * scaleY: 1 -> 0.5625
 */
const Grow = createStyleTransition(
  {
    enterStyle: () => ({
      opacity: 0,
      transform: getScale(0.75),
    }),
    enterActiveStyle: (instance, node) => ({
      transition: getEnterTransition(instance, node),
    }),
    enterToStyle: () => ({
      opacity: 1,
      transform: getScale(1),
    }),
    exitStyle: () => ({
      opacity: 1,
      transform: getScale(1),
    }),
    exitActiveStyle: (instance, node) => ({
      transition: getExitTransition(instance, node),
    }),
    exitToStyle: () => ({
      opacity: 0,
      transform: getScale(0.75),
    }),
  },
  defaultProps,
);

export default Grow;
