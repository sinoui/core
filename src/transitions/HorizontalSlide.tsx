import * as React from 'react';
import { transitions } from '@sinoui/theme';
import createStyleTransition from './createStyleTransition';
import { BaseTransitionCompProps } from './type';

const { duration } = transitions;

export interface HorizontalSlidePropsType extends BaseTransitionCompProps {
  /**
   * 过渡效果的方向
   */
  direction: 'left' | 'right';
  /**
   * 过渡时长。
   */
  timeout?:
    | number
    | {
        enter: number;
        exit: number;
      };
  /**
   * 过渡效果的缓动函数
   */
  easing?:
    | string
    | {
        enter: string;
        exit: string;
      };
}

// Slide效果默认的缓动函数
const SLIDE_EASING = {
  enter: 'cubic-bezier(0.23, 1, 0.32, 1)',
  exit: 'cubic-bezier(0.23, 1, 0.32, 1)',
};

// Slide效果默认的过渡时长
const TRANSITION_DURATION = {
  enter: duration.enteringScreen,
  exit: duration.leavingScreen,
};

/**
 * 水平滑动效果。内部使用
 * [react-transition-group/CSSTransition](https://reactcommunity.org/react-transition-group/#CSSTransition)实现过渡效果。
 */
const HorizontalSlide: React.ComponentType<
  HorizontalSlidePropsType | any
> = createStyleTransition({
  enterStyle: (transition) => ({
    transform: `translate(${
      transition.props.direction === 'right' ? '-' : ''
    }100%, 0px)`,
    opacity: 0.01,
  }),
  enterToStyle: () => ({
    transform: 'translateX(0px)',
    opacity: 1,
  }),
  exitStyle: () => ({
    opacity: 1,
    transform: 'translate3d(0px, 0px, 0px)',
  }),
  exitToStyle: (transition) => ({
    transform: `translate(${
      transition.props.direction === 'right' ? '-' : ''
    }100%, 0px)`,
    opacity: 0.01,
  }),
  transitionProps: ['transform', 'opacity'],
  easing: SLIDE_EASING,
  duration: TRANSITION_DURATION,
});

export default HorizontalSlide;
