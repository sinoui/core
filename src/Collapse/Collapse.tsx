/* eslint-disable no-param-reassign */
import React, { useRef, useEffect, useLayoutEffect } from 'react';
import { Transition } from 'react-transition-group';
import styled from 'styled-components';
import transitions from '@sinoui/theme/transitions';
import type { TransitionStatus } from 'react-transition-group/Transition';
import type BaseTransitionProps from '../transitions/BaseTransitionProps';
import getDuration from '../transitions/getDuration';
import type { TransitionTimeout } from '../transitions/getDuration';
import useMultiRefs from '../utils/useMultiRefs';

interface CollapseWrapperProps {
  $entered?: boolean;
  $hidden?: boolean;
  $minHeight: number;
}

const CollapseWrapper = styled.div<CollapseWrapperProps>`
  height: ${({ $entered }) => ($entered ? 'auto' : '0px')};
  overflow: hidden;
  transition: ${({ theme }) => theme.transitions.create('height')};
  visibility: ${({ $hidden }) => ($hidden ? 'hidden' : '')};
  min-height: ${({ $minHeight }) => $minHeight}px;

  & > div {
    display: flex;
    > div {
      width: 100%;
    }
  }
`;

type Props = Omit<BaseTransitionProps, 'timeout'> & {
  /**
   * 压缩后的高度。默认为`0`。单位为`px`。
   */
  collapsedHeight?: number;
  timeout?: TransitionTimeout;
};

const reflow = (node: HTMLElement) => node.offsetTop;

/**
 * 压缩元素的过渡效果。
 */
const Collapse = React.forwardRef<HTMLDivElement, Props>(function Collapse(
  props,
  ref,
) {
  const {
    children,
    timeout = transitions.duration.standard,
    in: inProp,
    collapsedHeight = 0,
    onEnter,
    onEntering,
    onEntered,
    onExit,
    onExiting,
    ...rest
  } = props;
  const contentRef = useRef<HTMLElement>(null);
  const handleContentRef = useMultiRefs(contentRef, ref);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const autoTimeout = useRef<number>(0);
  const timerRef = useRef<number>();

  const handleEnter: any = (isAppearing: boolean) => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    node.style.height = `${collapsedHeight}px`;
    if (onEnter) {
      onEnter(node, isAppearing);
    }
  };

  const handleEntering: any = (isAppearing: boolean) => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    const wrapper = wrapperRef.current;
    const duration = getDuration(timeout, 'enter', wrapper);
    node.style.transitionDuration = `${duration}ms`;
    node.style.height = `${wrapper?.clientHeight ?? 0}px`;
    autoTimeout.current = timeout === 'auto' ? duration : 0;

    if (onEntering) {
      onEntering(node, isAppearing);
    }
  };

  const handleEntered: any = (isAppearing: boolean) => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    node.style.height = 'auto';
    if (onEntered) {
      onEntered(node, isAppearing);
    }
  };

  const handleExit = () => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    node.style.height = `${wrapperRef.current?.clientHeight ?? 0}px`;
    reflow(node);
    if (onExit) {
      onExit(node);
    }
  };

  const handleExiting = () => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    const wrapper = wrapperRef.current;
    const duration = getDuration(timeout, 'exit', wrapper);
    node.style.transitionDuration = `${duration}ms`;
    autoTimeout.current = timeout === 'auto' ? duration : 0;
    node.style.height = `${collapsedHeight}px`;

    if (onExiting) {
      onExiting(node);
    }
  };

  const addEndListener: any = (done: () => void) => {
    timerRef.current = setTimeout(done, autoTimeout.current);
  };

  useLayoutEffect(() => {
    return () => window.clearTimeout(timerRef.current);
  }, []);

  return (
    <Transition
      appear
      in={inProp}
      timeout={timeout === 'auto' ? undefined : timeout}
      onEnter={handleEnter}
      onEntering={handleEntering}
      onEntered={handleEntered}
      onExit={handleExit}
      onExiting={handleExiting}
      addEndListener={(timeout === 'auto' ? addEndListener : undefined) as any}
      nodeRef={wrapperRef}
      {...rest}
    >
      {(status: TransitionStatus, childProps: any) => (
        <CollapseWrapper
          $entered={status === 'entered'}
          $hidden={status === 'exited' && !inProp && collapsedHeight === 0}
          $minHeight={collapsedHeight}
          ref={handleContentRef}
          {...childProps}
        >
          <div ref={wrapperRef}>
            <div>{children}</div>
          </div>
        </CollapseWrapper>
      )}
    </Transition>
  );
});

export default Collapse;
