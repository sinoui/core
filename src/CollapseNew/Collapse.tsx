/* eslint-disable no-param-reassign */
import React, { useRef, useEffect, useLayoutEffect } from 'react';
import { Transition } from 'react-transition-group';
import styled from 'styled-components';
import transitions from '@sinoui/theme/transitions';
import type { TransitionStatus } from 'react-transition-group/Transition';
import { animate } from '@sinoui/utils';
import type BaseTransitionProps from '../transitions/BaseTransitionProps';
import getDuration from '../transitions/getDuration';
import type { TransitionTimeout } from '../transitions/getDuration';
import useMultiRefs from '../utils/useMultiRefs';

interface CollapseWrapperProps {
  $hidden?: boolean;
}

const CollapseWrapper = styled.div<CollapseWrapperProps>`
  overflow: hidden;
  visibility: ${({ $hidden }) => ($hidden ? 'hidden' : '')};
  transform-origin: top left;

  & > div {
    display: flex;
    transform-origin: top left;
    will-change: transform;
  }
`;

type Props = Omit<BaseTransitionProps, 'timeout'> & {
  /**
   * 压缩后的高度。默认为`0`。单位为`px`。
   */
  collapsedHeight?: number;
  timeout?: TransitionTimeout;
  /**
   * 收缩方向，auto表示垂直和水平方向同时收缩
   */
  direction?: 'vertical' | 'horizontal' | 'auto';
};

/**
 * 压缩元素的过渡效果。
 */
const Collapse = React.forwardRef<HTMLDivElement, Props>(function Collapse(
  props,
  ref,
) {
  const {
    children,
    timeout = transitions.duration.standard,
    in: inProp,
    collapsedHeight = 0,
    onEnter,
    onEntering,
    onEntered,
    onExit,
    onExiting,
    direction = 'vertical',
    ...rest
  } = props;
  const contentRef = useRef<HTMLElement>(null);
  const handleContentRef = useMultiRefs(contentRef, ref);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const autoTimeout = useRef<number>(0);
  const timerRef = useRef<number>();
  const enterAnimateRef = useRef<any>(null);
  const exitAnimateRef = useRef<any>(null);
  const inRef = useRef<boolean>(!!inProp);

  const handleEnter: any = (isAppearing: boolean) => {
    const node = contentRef.current;
    const content = wrapperRef.current;
    if (!node || !content) {
      return;
    }
    const start = collapsedHeight / content.clientHeight;

    enterAnimateRef.current = animate(start, 1, 200, (value) => {
      const scaleX = direction === 'vertical' ? 1 : value;
      const scaleY = direction === 'horizontal' ? 1 : value;
      const invScaleX = scaleX === 0 ? 60 : 1 / scaleX;
      const invScalY = scaleY === 0 ? 60 : 1 / scaleY;
      node.style.transform = `scale(${scaleX}, ${scaleY})`;
      content.style.transform = `scale(${invScaleX}, ${invScalY})`;
    });

    if (onEnter) {
      onEnter(node, isAppearing);
    }
  };

  const handleEntering: any = (isAppearing: boolean) => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    const wrapper = wrapperRef.current;
    const duration = getDuration(timeout, 'enter', wrapper);
    autoTimeout.current = timeout === 'auto' ? duration : 0;

    if (onEntering) {
      onEntering(node, isAppearing);
    }
  };

  const handleEntered: any = (isAppearing: boolean) => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    enterAnimateRef.current();
    if (onEntered) {
      onEntered(node, isAppearing);
    }
  };

  const handleExit = () => {
    const node = contentRef.current;
    const content = wrapperRef.current;
    if (!node || !content) {
      return;
    }
    const end = collapsedHeight / content.clientHeight;
    exitAnimateRef.current = animate(1, end, 200, (value) => {
      const scaleX = direction === 'vertical' ? 1 : value;
      const scaleY = direction === 'horizontal' ? 1 : value;
      const invScaleX = scaleX === 0 ? 60 : 1 / scaleX;
      const invScalY = scaleY === 0 ? 60 : 1 / scaleY;
      node.style.transform = `scale(${scaleX}, ${scaleY})`;
      content.style.transform = `scale(${invScaleX}, ${invScalY})`;
    });

    if (onExit) {
      onExit(node);
    }
  };

  const handleExiting = () => {
    const node = contentRef.current;
    if (!node) {
      return;
    }
    const wrapper = wrapperRef.current;
    const duration = getDuration(timeout, 'exit', wrapper);

    autoTimeout.current = timeout === 'auto' ? duration : 0;

    if (onExiting) {
      onExiting(node);
    }
  };

  const handleExited = () => {
    exitAnimateRef.current();
  };

  const addEndListener: any = (done: () => void) => {
    timerRef.current = setTimeout(done, autoTimeout.current);
  };

  useEffect(() => {
    return () => window.clearTimeout(timerRef.current);
  }, []);

  useLayoutEffect(() => {
    if (!inRef.current) {
      const node = contentRef.current;
      const content = wrapperRef.current;
      if (!node || !content) {
        return;
      }

      const scaleY = collapsedHeight / content.clientHeight;
      const invScalY = scaleY === 0 ? 60 : 1 / scaleY;

      node.style.transform = `scale(1,${scaleY})`;
      content.style.transform = `scale(1,${invScalY})`;
    }
  }, [collapsedHeight]);

  return (
    <Transition
      appear
      in={inProp}
      timeout={timeout === 'auto' ? undefined : timeout}
      onEnter={handleEnter}
      onEntering={handleEntering}
      onEntered={handleEntered}
      onExit={handleExit}
      onExiting={handleExiting}
      onExited={handleExited}
      addEndListener={(timeout === 'auto' ? addEndListener : undefined) as any}
      nodeRef={wrapperRef}
      {...rest}
    >
      {(status: TransitionStatus, childProps: any) => (
        <CollapseWrapper
          className="sinoui-collapse-container"
          $hidden={status === 'exited' && !inProp && collapsedHeight === 0}
          ref={handleContentRef}
          {...childProps}
        >
          <div ref={wrapperRef} className="sinoui-collapse-content">
            {children}
          </div>
        </CollapseWrapper>
      )}
    </Transition>
  );
});

export default Collapse;
