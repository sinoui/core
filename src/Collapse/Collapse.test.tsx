import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { duration } from '@sinoui/theme/transitions';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import 'jest-styled-components';
import { create } from 'react-test-renderer';
import Collapse from './Collapse';

afterEach(cleanup);

jest.useFakeTimers();

const handleRef = (node: HTMLDivElement) => {
  if (node) {
    Object.defineProperty(node.firstChild, 'clientHeight', {
      configurable: true,
      value: 200,
    });
  }
};

it('默认展现', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        data-testid="collapse"
        in
        ref={(node) => {
          if (node) {
            Object.defineProperty(node.firstChild, 'clientHeight', {
              configurable: true,
              value: 200,
            });
          }
        }}
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle(
    `height: 200px; transition-duration: ${duration.standard}ms;`,
  );

  // onEntered: `height: auto`
  jest.runAllTimers();
  expect(getByTestId('collapse')).toHaveStyle('height: auto');
});

it('默认隐藏', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse data-testid="collapse" in={false}>
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyleRule('min-height', '0px');
  expect(getByTestId('collapse')).toHaveStyleRule('height', '0px');
});

it('collapseHeight', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse data-testid="collapse" in={false} collapsedHeight={20}>
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyleRule('min-height', '20px');
});

it('显示 -> 隐藏', () => {
  const { getByTestId, rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse data-testid="collapse" in ref={handleRef}>
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  let heightOnExit = '0px';
  rerender(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        data-testid="collapse"
        in={false}
        ref={handleRef}
        onExit={(node) => {
          heightOnExit = node.style.height;
        }}
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(heightOnExit).toBe('200px');
  expect(getByTestId('collapse')).toHaveStyle('height: 0px');
  expect(getByTestId('collapse')).toHaveStyle(
    `transition-duration: ${duration.standard}ms`,
  );
});

it('隐藏 -> 显示', () => {
  const { getByTestId, rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        data-testid="collapse"
        in={false}
        ref={handleRef}
        collapsedHeight={20}
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  let heightOnEnter = '0px';
  rerender(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        data-testid="collapse"
        in
        ref={handleRef}
        collapsedHeight={20}
        onEnter={(node) => {
          heightOnEnter = node.style.height;
        }}
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  // onEnter: height: 20px
  expect(heightOnEnter).toBe('20px');
  // onEntering: height: 200px
  expect(getByTestId('collapse')).toHaveStyle('height: 200px');
  // onEntered: `height: auto`
  jest.runAllTimers();
  expect(getByTestId('collapse')).toHaveStyle('height: auto');
});

it('auto timeout', () => {
  const { getByTestId, rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse in timeout="auto" ref={handleRef} data-testid="collapse">
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle('transition-duration: 281ms');

  // 验证addEndListener
  jest.runAllTimers();
  expect(getByTestId('collapse')).toHaveStyle('height: auto');

  rerender(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        in={false}
        timeout="auto"
        ref={handleRef}
        data-testid="collapse"
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle('transition-duration: 281ms');
});

it('clean timeout after unmounted', () => {
  const onEntered = jest.fn();
  const { unmount } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        in
        timeout="auto"
        ref={handleRef}
        data-testid="collapse"
        onEntered={onEntered}
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  unmount();
  jest.runAllTimers();
  expect(onEntered).not.toBeCalled();
});

it('ref', () => {
  const ref = React.createRef<HTMLDivElement>();
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse in timeout="auto" data-testid="collapse" ref={ref}>
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(ref.current).toBe(getByTestId('collapse'));
});

it('onEnter, onEntering, onEntered', () => {
  const onEnter = jest.fn();
  const onEntering = jest.fn();
  const onEntered = jest.fn();

  render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse
        in
        onEnter={onEnter}
        onEntering={onEntering}
        onEntered={onEntered}
      >
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(onEnter).toBeCalled();
  expect(onEntering).toBeCalled();
  expect(onEntered).not.toBeCalled();

  jest.runAllTimers();

  expect(onEntered).toBeCalled();
});

it('onExit, onExiting', () => {
  const onExit = jest.fn();
  const onExiting = jest.fn();
  const { rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <Collapse in onExit={onExit} onExiting={onExiting}>
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );
  rerender(
    <ThemeProvider theme={defaultTheme}>
      <Collapse in={false} onExit={onExit} onExiting={onExiting}>
        <div>测试</div>
      </Collapse>
    </ThemeProvider>,
  );

  expect(onExit).toBeCalled();
  expect(onExiting).toBeCalled();
});

describe('快照测试', () => {
  it('展现', () => {
    const tree = create(
      <React.StrictMode>
        <ThemeProvider theme={defaultTheme}>
          <Collapse in>
            <div>测试</div>
          </Collapse>
        </ThemeProvider>
      </React.StrictMode>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
