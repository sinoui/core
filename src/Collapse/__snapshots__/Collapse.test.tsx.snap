// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 展现 1`] = `
.c0 {
  height: 0px;
  overflow: hidden;
  -webkit-transition: height 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: height 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  min-height: 0px;
}

.c0 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 > div > div {
  width: 100%;
}

<div
  className="c0"
>
  <div>
    <div>
      <div>
        测试
      </div>
    </div>
  </div>
</div>
`;
